crates/subspace-networking/src/peer_info/protocol.rs

```diff
@@ -0,0 +1,39 @@
+//! This module defines low-level functions for working with inbound and outbound streams.
+
+use crate::peer_info::PeerInfo;
+use futures::prelude::*;
+use parity_scale_codec::{Decode, Encode};
+use std::io;
+use std::io::ErrorKind;
+
+/// Send peer-info data to a remote peer.
+pub async fn send<S>(mut stream: S, pi: PeerInfo) -> io::Result<S>
+where
+    S: AsyncRead + AsyncWrite + Unpin,
+{
+    let send_data = pi.encode();
+    let send_len_bytes = (send_data.len() as u32).to_le_bytes();
+
+    stream.write_all(&send_len_bytes).await?;
+    stream.write_all(&send_data).await?;
+    stream.flush().await?;
+
+    Ok(stream)
+}
+
+/// Receive peer-info data from a remote peer.
+pub async fn recv<S>(mut stream: S) -> io::Result<(S, PeerInfo)>
+where
+    S: AsyncRead + AsyncWrite + Unpin,
+{
+    let mut rec_len_bytes = [0u8; 4];
```

Comment:
There is nothing wrong here, but I typically prefer writing it in more explicit way:
```
    let mut rec_len_bytes = 0u32.to_le_bytes();
```

Then it will be symmetric to `from_le_bytes()` below.